tests/test_grad.py

```diff
@@ -0,0 +1,403 @@
+    sens_type = {'type': sens_type}
+    steps = ['resolve', 'infer', 'specialize', 'prepare',
+             'validate', 'export']
+    if isinstance(obj, FunctionType):
+        res = standard_debug_pipeline \
+            .select('parse', *steps) \
+            .insert_before('infer', wrap=grad_wrap) \
+            .insert_after('prepare', opt=step_optgrad) \
+            .run(input=obj, argspec=[*in_types, sens_type])
+    else:
+        res = standard_debug_pipeline \
+            .select(*steps) \
+            .insert_before(wrap=grad_wrap) \
+            .insert_after('prepare', opt=step_optgrad) \
+            .run(graph=obj, argspec=[*in_types, sens_type])
+
+    gtest = GradTester(
+        fn=fn,
+        gfn=res['output'],
+        args=args,
+        argnames=[f'in{i}' for i in range(len(args))],
+        outnames=None
+    )
+    gtest.assert_match()
+
+
+@pytest.mark.parametrize('prim,cases', prim_tests.items())
+def test_prim_grads(prim, cases):
+    for case in cases:
+        _grad_test(pyi[prim], prim, case)
+
+
+def grad_test(*tests):
+    """Decorate a function to parse and run it against pure Python.
+
+    Returns a unit test that will parse the function, and then for
+    each `inputs` tuple in `tests` it will check that the pure Python,
+    undecorated function returns that same output.
+
+    Arguments:
+        tests: One or more inputs tuple.
+
+    """
+
+    def decorate(fn):
+        def test(args):
+            if not isinstance(args, tuple):
+                args = (args,)
+
+            _grad_test(fn, fn, args)
+
+        m = pytest.mark.parametrize('args', list(tests))(test)
+        m.__orig__ = fn
+        return m
+    return decorate
+
+
+@grad_test((13.0, 14.0))
+def test_null(x, y):
+    """Test null gradient."""
+    return 10.0 + 28.0 / 43.0
+
+
+@grad_test((1.0, 4.0), (5.0, -13.0))
+def test_grad_add(x, y):
+    return x + y
+
+
+@grad_test((3.0, 4.0))
+def test_grad_expr(x, y):
+    return x**3.0 * y**4.0
+
+
+@grad_test((3.0,))
+def test_constant(x):
+    """Test the use of a literal in the expression."""
+    return 18.0 * x
+
+
+@grad_test((3.0,))
+def test_dup_args_in_call(x):
+    """The naive gradient update rule fails when a function's arguments
+    contain the same variable more than once."""
+    return x * x
+
+
+@grad_test((3.0,))
+def test_quadruple_args_in_call(x):
+    """Test that duplicated arguments still cause no problem even if
+    there are four of them."""
+    def g(a, b, c, d):
+        return a * b * c * d
+    return g(x, x, x, x)
+
+
+@grad_test((3.0, 5.0))
+def test_tuples(x, y):
+    tup = x + y, x * y
+    z = tup[0] + tup[1]
+    return z
+
+
+@grad_test((4.0, 5.0))
+def test_hof(a, b):
+    """Test higher order functions."""
+    def f(g, x):
+        return g(x) * g(x + 10.0)
+
+    def g(x):
+        return x * b
+
+    return f(g, a) + f(g, b)
+
+
+@pytest.mark.xfail(
+    reason="I think the opts to eliminate j/jinv are messing this one up."
```

Comment:
If this is true then the opts should be fixed or disabled.

Reply:
I disabled the opts and added support for J/Jinv in py_implementations.